src/model/id.rs

```diff
@@ -1,338 +1,315 @@
 //! A collection of newtypes defining type-strong IDs.
+// These warns fire inside derivative
+#![allow(
+    clippy::non_canonical_partial_ord_impl,
+    clippy::non_canonical_clone_impl,
+    clippy::let_underscore_must_use
+)]
 
 use std::fmt;
+use std::marker::PhantomData;
 use std::num::{NonZeroI64, NonZeroU64};
 
 use super::Timestamp;
 
-macro_rules! id_u64 {
-    ($($name:ident;)*) => {
-        $(
-            impl $name {
-                /// Creates a new Id from a u64
-                ///
-                /// # Panics
-                /// Panics if the id is zero.
-                #[inline]
-                #[must_use]
-                #[track_caller]
-                pub const fn new(id: u64) -> Self {
-                    match NonZeroU64::new(id) {
-                        Some(inner) => Self(inner),
-                        None => panic!("Attempted to call Id::new with invalid (0) value")
-                    }
-                }
-
-                /// Retrieves the inner ID as u64
-                #[inline]
-                #[must_use]
-                pub const fn get(self) -> u64 {
-                    self.0.get()
-                }
-
-                /// Retrieves the time that the Id was created at.
-                #[must_use]
-                pub fn created_at(&self) -> Timestamp {
-                    Timestamp::from_discord_id(self.get())
-                }
-
-                /// Immutably borrow inner Id.
-                #[inline]
-                #[must_use]
-                pub const fn as_inner(&self) -> &NonZeroU64 {
-                    &self.0
-                }
-            }
+#[derive(derivative::Derivative)]
+#[derivative(
+    Clone(bound = ""),
+    Copy(bound = ""),
+    Debug(bound = ""),
+    Eq(bound = ""),
+    Hash(bound = ""),
+    Ord(bound = ""),
+    PartialEq(bound = ""),
+    PartialOrd(bound = "")
+)]
```

Comment:
What's the reason for using the derivative crate?

Reply:
If it didn't, the Marker types would have to be all of these traits for any of the derives to work. derivative allows providing the bounds (as none) to avoid having to write all the implementations out manually.

Reply:
I think this could be achieved by just manually implementing the traits for `Id`. Additionally, I think it would be useful to have an `IdMarker` trait that acts as a marker trait (heh) for `Id` to bound its parameter over, as having `Id` be completely generic doesn't really make sense.

Reply:
Tried my hand at it and this is what I got. The extra variable assignments are due to the fact that making a reference to a packed field can potentially cause undefined behavior if the reference is unaligned. Therefore, copying the value onto the stack and passing a reference to the copy avoids this problem.

```rust
impl<Marker> Copy for Id<Marker> {}
impl<Marker> Clone for Id<Marker> {
    fn clone(&self) -> Self {
        *self
    }
}

impl<Marker> fmt::Debug for Id<Marker> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let inner = self.inner;
        write!(f, "{:?}", inner)
    }
}

impl<Marker> PartialEq for Id<Marker> {
    fn eq(&self, other: &Self) -> bool {
        let inner = self.inner;
        let other_inner = other.inner;
        inner == other_inner
    }
}

impl<Marker> Eq for Id<Marker> {}

impl<Marker> Hash for Id<Marker> {
    fn hash<H: Hasher>(&self, state: &mut H) {
        let inner = self.inner;
        inner.hash(state)
    }
}

impl<Marker> PartialOrd for Id<Marker> {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        let inner = self.inner;
        let other_inner = other.inner;
        inner.partial_cmp(&other_inner)
    }
}

impl<Marker> Ord for Id<Marker> {
    fn cmp(&self, other: &Self) -> Ordering {
        let inner = self.inner;
        let other_inner = other.inner;
        inner.cmp(&other_inner)
    }
}
```

Reply:
I never said they couldn't be implemented, I just used derivative because it's easier and I didn't feel like writing them all out as you did. If this is a blocker to merge though, I'll be fine to copy paste that into this PR.

Reply:
This approach is lighter weight and avoids the compile-time hit of adding `derivative`, which wasn't anywhere in the dependency tree before.